// @designliquido/visualg/lexador/palavras-reservadas.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.palavrasReservadas = void 0;
const visualg_1 = __importDefault(require("@designliquido/delegua/tipos-de-simbolos/visualg"));
exports.palavrasReservadas = {
    abs: visualg_1.default.METODO_BIBLIOTECA_GLOBAL,
    algoritmo: visualg_1.default.ALGORITMO,
    aleatorio: visualg_1.default.ALEATORIO,
    arccos: visualg_1.default.METODO_BIBLIOTECA_GLOBAL,
    arcsen: visualg_1.default.METODO_BIBLIOTECA_GLOBAL,
    arctan: visualg_1.default.METODO_BIBLIOTECA_GLOBAL,
    ate: visualg_1.default.ATE,
    até: visualg_1.default.ATÉ,
    caracter: visualg_1.default.CARACTER,
    caractere: visualg_1.default.CARACTERE,
    caso: visualg_1.default.CASO,
    cos: visualg_1.default.METODO_BIBLIOTECA_GLOBAL,
    cotan: visualg_1.default.METODO_BIBLIOTECA_GLOBAL,
    de: visualg_1.default.DE,
    e: visualg_1.default.E,
    enquanto: visualg_1.default.ENQUANTO,
    entao: visualg_1.default.ENTAO,
    então: visualg_1.default.ENTÃO,
    escolha: visualg_1.default.ESCOLHA,
    escreva: visualg_1.default.ESCREVA,
    escreval: visualg_1.default.ESCREVA_LINHA,
    exp: visualg_1.default.METODO_BIBLIOTECA_GLOBAL,
    faca: visualg_1.default.FACA,
    faça: visualg_1.default.FAÇA,
    falso: visualg_1.default.FALSO,
    fimalgoritmo: visualg_1.default.FIM_ALGORITMO,
    fimenquanto: visualg_1.default.FIM_ENQUANTO,
    fimescolha: visualg_1.default.FIM_ESCOLHA,
    fimfuncao: visualg_1.default.FIM_FUNCAO,
    fimfunção: visualg_1.default.FIM_FUNÇÃO,
    fimpara: visualg_1.default.FIM_PARA,
    fimprocedimento: visualg_1.default.FIM_PROCEDIMENTO,
    fimrepita: visualg_1.default.FIM_REPITA,
    fimse: visualg_1.default.FIM_SE,
    funcao: visualg_1.default.FUNCAO,
    função: visualg_1.default.FUNÇÃO,
    grauprad: visualg_1.default.METODO_BIBLIOTECA_GLOBAL,
    inicio: visualg_1.default.INICIO,
    int: visualg_1.default.METODO_BIBLIOTECA_GLOBAL,
    inteiro: visualg_1.default.INTEIRO,
    interrompa: visualg_1.default.INTERROMPA,
    leia: visualg_1.default.LEIA,
    limpatela: visualg_1.default.METODO_BIBLIOTECA_GLOBAL,
    log: visualg_1.default.METODO_BIBLIOTECA_GLOBAL,
    logn: visualg_1.default.METODO_BIBLIOTECA_GLOBAL,
    logico: visualg_1.default.LOGICO,
    nao: visualg_1.default.NEGACAO,
    on: visualg_1.default.ON,
    off: visualg_1.default.OFF,
    ou: visualg_1.default.OU,
    outrocaso: visualg_1.default.OUTRO_CASO,
    para: visualg_1.default.PARA,
    passo: visualg_1.default.PASSO,
    pi: visualg_1.default.METODO_BIBLIOTECA_GLOBAL,
    procedimento: visualg_1.default.PROCEDIMENTO,
    quad: visualg_1.default.METODO_BIBLIOTECA_GLOBAL,
    radpgrau: visualg_1.default.METODO_BIBLIOTECA_GLOBAL,
    raizq: visualg_1.default.METODO_BIBLIOTECA_GLOBAL,
    rand: visualg_1.default.METODO_BIBLIOTECA_GLOBAL,
    randi: visualg_1.default.METODO_BIBLIOTECA_GLOBAL,
    real: visualg_1.default.REAL,
    repita: visualg_1.default.REPITA,
    retorne: visualg_1.default.RETORNE,
    se: visualg_1.default.SE,
    sen: visualg_1.default.METODO_BIBLIOTECA_GLOBAL,
    senao: visualg_1.default.SENAO,
    senão: visualg_1.default.SENÃO,
    tan: visualg_1.default.METODO_BIBLIOTECA_GLOBAL,
    var: visualg_1.default.VAR,
    verdadeiro: visualg_1.default.VERDADEIRO,
    vetor: visualg_1.default.VETOR,
    xou: visualg_1.default.XOU,
};
