// @designliquido/delegua/estruturas/delegua-classe.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DeleguaClasse = void 0;
const excecoes_1 = require("../excecoes");
const chamavel_1 = require("./chamavel");
const objeto_delegua_classe_1 = require("./objeto-delegua-classe");
/**
 * Estrutura de declaração de classe.
 */
class DeleguaClasse extends chamavel_1.Chamavel {
    constructor(simboloOriginal, superClasse, metodos, propriedades) {
        super();
        this.simboloOriginal = simboloOriginal;
        this.superClasse = superClasse;
        this.metodos = metodos || {};
        this.propriedades = propriedades || [];
        this.dialetoRequerDeclaracaoPropriedades = false;
    }
    encontrarMetodo(nome) {
        if (this.metodos.hasOwnProperty(nome)) {
            return this.metodos[nome];
        }
        if (this.superClasse !== null && this.superClasse !== undefined) {
            return this.superClasse.encontrarMetodo(nome);
        }
        return undefined;
    }
    encontrarPropriedade(nome) {
        if (nome in this.propriedades) {
            return this.propriedades[nome];
        }
        if (this.superClasse !== null && this.superClasse !== undefined) {
            return this.superClasse.encontrarPropriedade(nome);
        }
        if (this.dialetoRequerDeclaracaoPropriedades) {
            throw new excecoes_1.ErroEmTempoDeExecucao(this.simboloOriginal, `Propriedade "${nome}" não declarada na classe ${this.simboloOriginal.lexema}.`);
        }
        return undefined;
    }
    /**
     * Método utilizado por Delégua para representar esta classe quando impressa.
     * @returns {string} A representação da classe como texto.
     */
    paraTexto() {
        let texto = `<classe ${this.simboloOriginal.lexema}`;
        for (let propriedade of this.propriedades) {
            texto += ` ${propriedade.nome}`;
            if (propriedade.tipo) {
                texto += `:${propriedade.tipo}`;
            }
            texto += ' ';
        }
        texto += '>';
        return texto;
    }
    /**
     * Método utilizado pelo VSCode para inspecionar esta classe em depuração.
     * @returns {string} A representação da classe como texto.
     */
    toString() {
        return this.paraTexto();
    }
    aridade() {
        const inicializador = this.encontrarMetodo('construtor');
        return inicializador ? inicializador.aridade() : 0;
    }
    async chamar(visitante, argumentos) {
        const instancia = new objeto_delegua_classe_1.ObjetoDeleguaClasse(this);
        const inicializador = this.encontrarMetodo('construtor');
        if (inicializador) {
            const metodoConstrutor = inicializador.funcaoPorMetodoDeClasse(instancia);
            await metodoConstrutor.chamar(visitante, argumentos);
        }
        return instancia;
    }
}
exports.DeleguaClasse = DeleguaClasse;
