// @designliquido/delegua/lexador/palavras-reservadas.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.palavrasReservadasMicroGramatica = exports.palavrasReservadas = void 0;
const delegua_1 = __importDefault(require("../tipos-de-simbolos/delegua"));
exports.palavrasReservadas = {
    cada: delegua_1.default.CADA,
    caso: delegua_1.default.CASO,
    classe: delegua_1.default.CLASSE,
    como: delegua_1.default.COMO,
    construtor: delegua_1.default.CONSTRUTOR,
    continua: delegua_1.default.CONTINUA,
    constante: delegua_1.default.CONSTANTE,
    const: delegua_1.default.CONSTANTE,
    de: delegua_1.default.DE,
    e: delegua_1.default.E,
    em: delegua_1.default.EM,
    enquanto: delegua_1.default.ENQUANTO,
    escolha: delegua_1.default.ESCOLHA,
    escreva: delegua_1.default.ESCREVA,
    falhar: delegua_1.default.FALHAR,
    falso: delegua_1.default.FALSO,
    fazer: delegua_1.default.FAZER,
    finalmente: delegua_1.default.FINALMENTE,
    fixo: delegua_1.default.CONSTANTE,
    funcao: delegua_1.default.FUNCAO,
    função: delegua_1.default.FUNÇÃO,
    herda: delegua_1.default.HERDA,
    importar: delegua_1.default.IMPORTAR,
    isto: delegua_1.default.ISTO,
    leia: delegua_1.default.LEIA,
    nulo: delegua_1.default.NULO,
    ou: delegua_1.default.OU,
    padrao: delegua_1.default.PADRAO,
    para: delegua_1.default.PARA,
    pausa: delegua_1.default.PAUSA,
    pegue: delegua_1.default.PEGUE,
    retorna: delegua_1.default.RETORNA,
    se: delegua_1.default.SE,
    senao: delegua_1.default.SENAO,
    senão: delegua_1.default.SENÃO,
    super: delegua_1.default.SUPER,
    sustar: delegua_1.default.SUSTAR,
    tendo: delegua_1.default.TENDO,
    tente: delegua_1.default.TENTE,
    tipo: delegua_1.default.TIPO,
    var: delegua_1.default.VARIAVEL,
    variavel: delegua_1.default.VARIAVEL,
    variável: delegua_1.default.VARIAVEL,
    verdadeiro: delegua_1.default.VERDADEIRO,
};
exports.palavrasReservadasMicroGramatica = {
    e: delegua_1.default.E,
    em: delegua_1.default.EM,
    falso: delegua_1.default.FALSO,
    nulo: delegua_1.default.NULO,
    ou: delegua_1.default.OU,
    verdadeiro: delegua_1.default.VERDADEIRO,
};
