// @designliquido/delegua/lexador/dialetos/palavras-reservadas/egua-classico.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.palavrasReservadas = void 0;
const egua_classico_1 = __importDefault(require("../../../tipos-de-simbolos/egua-classico"));
exports.palavrasReservadas = {
    e: egua_classico_1.default.E,
    em: egua_classico_1.default.EM,
    caso: egua_classico_1.default.CASO,
    classe: egua_classico_1.default.CLASSE,
    continua: egua_classico_1.default.CONTINUA,
    enquanto: egua_classico_1.default.ENQUANTO,
    escolha: egua_classico_1.default.ESCOLHA,
    escreva: egua_classico_1.default.ESCREVA,
    falso: egua_classico_1.default.FALSO,
    fazer: egua_classico_1.default.FAZER,
    finalmente: egua_classico_1.default.FINALMENTE,
    função: egua_classico_1.default.FUNCAO,
    herda: egua_classico_1.default.HERDA,
    importar: egua_classico_1.default.IMPORTAR,
    isto: egua_classico_1.default.ISTO,
    leia: egua_classico_1.default.LEIA,
    nulo: egua_classico_1.default.NULO,
    ou: egua_classico_1.default.OU,
    para: egua_classico_1.default.PARA,
    padrão: egua_classico_1.default.PADRAO,
    pausa: egua_classico_1.default.PAUSA,
    pegue: egua_classico_1.default.PEGUE,
    retorna: egua_classico_1.default.RETORNA,
    se: egua_classico_1.default.SE,
    senãose: egua_classico_1.default.SENÃOSE,
    senão: egua_classico_1.default.SENÃO,
    super: egua_classico_1.default.SUPER,
    tente: egua_classico_1.default.TENTE,
    var: egua_classico_1.default.VARIAVEL,
    verdadeiro: egua_classico_1.default.VERDADEIRO,
};
