// @designliquido/delegua/lexador/dialetos/palavras-reservadas/potigol.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.palavrasReservadas = void 0;
const potigol_1 = __importDefault(require("../../../tipos-de-simbolos/potigol"));
exports.palavrasReservadas = {
    ate: potigol_1.default.ATE,
    até: potigol_1.default.ATE,
    Caractere: potigol_1.default.CARACTERE,
    caso: potigol_1.default.CASO,
    de: potigol_1.default.DE,
    div: potigol_1.default.DIVISAO_INTEIRA,
    e: potigol_1.default.E,
    escolha: potigol_1.default.ESCOLHA,
    entao: potigol_1.default.ENTAO,
    então: potigol_1.default.ENTAO,
    enquanto: potigol_1.default.ENQUANTO,
    escreva: potigol_1.default.ESCREVA,
    faca: potigol_1.default.FACA,
    faça: potigol_1.default.FACA,
    falso: potigol_1.default.FALSO,
    fim: potigol_1.default.FIM,
    imprima: potigol_1.default.IMPRIMA,
    Inteiro: potigol_1.default.INTEIRO,
    leia_inteiro: potigol_1.default.LEIA_INTEIRO,
    leia_inteiros: potigol_1.default.LEIA_INTEIROS,
    leia_reais: potigol_1.default.LEIA_REAIS,
    leia_real: potigol_1.default.LEIA_REAL,
    leia_texto: potigol_1.default.LEIA_TEXTO,
    leia_textos: potigol_1.default.LEIA_TEXTOS,
    Logico: potigol_1.default.LOGICO,
    Lógico: potigol_1.default.LOGICO,
    mod: potigol_1.default.MODULO,
    ou: potigol_1.default.OU,
    nao: potigol_1.default.NEGACAO,
    não: potigol_1.default.NEGACAO,
    para: potigol_1.default.PARA,
    passo: potigol_1.default.PASSO,
    Real: potigol_1.default.REAL,
    retorne: potigol_1.default.RETORNE,
    se: potigol_1.default.SE,
    senao: potigol_1.default.SENAO,
    senão: potigol_1.default.SENAO,
    Texto: potigol_1.default.TEXTO,
    tipo: potigol_1.default.TIPO,
    var: potigol_1.default.VARIAVEL,
    verdadeiro: potigol_1.default.VERDADEIRO,
    qual_tipo: potigol_1.default.QUAL_TIPO,
};
