// @designliquido/delegua/lexador/dialetos/palavras-reservadas/birl.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.palavrasReservadas = void 0;
const birl_1 = __importDefault(require("../../../tipos-de-simbolos/birl"));
exports.palavrasReservadas = {
    // HORA_DO_SHOW: tiposDeSimbolos.HORA_DO_SHOW,
    // BIRL: tiposDeSimbolos.BIRL,
    // QUE_QUE_CE_QUER_MONSTRAO: tiposDeSimbolos.QUE_QUE_CE_QUER_MONSTRAO,
    // ELE_QUE_A_GENTE_QUER: tiposDeSimbolos.ELE_QUE_A_GENTE_QUER,
    // NAO_VAI_DAR_NAO: tiposDeSimbolos.NAO_VAI_DAR_NAO,
    // QUE_NAO_VAI_DAR_O_QUE: tiposDeSimbolos.QUE_NAO_VAI_DAR_O_QUE,
    // NEGATIVA_BAMBAM: tiposDeSimbolos.NEGATIVA_BAMBAM,
    // MAIS_QUERO_MAIS: tiposDeSimbolos.MAIS_QUERO_MAIS,
    // VAMO_MONSTRO: tiposDeSimbolos.VAMO_MONSTRO,
    // SAI_FILHO_DA_PUTA: tiposDeSimbolos.SAI_FILHO_DA_PUTA,
    // OH_O_HOME_AI_PO: tiposDeSimbolos.OH_O_HOME_AI_PO,
    // AJUDA_O_MALUCO_TA_DOENTE: tiposDeSimbolos.AJUDA_O_MALUCO_TA_DOENTE,
    a: birl_1.default.A,
    ai: birl_1.default.AI,
    ajuda: birl_1.default.AJUDA,
    bambam: birl_1.default.BAMBAM,
    birl: birl_1.default.BIRL,
    bora: birl_1.default.BORA,
    biceps: birl_1.default.BICEPS,
    ce: birl_1.default.CE,
    cumpade: birl_1.default.CUMPADE,
    da: birl_1.default.DA,
    dar: birl_1.default.DAR,
    do: birl_1.default.DO,
    doente: birl_1.default.DOENTE,
    descendente: birl_1.default.DESCENDENTE,
    ele: birl_1.default.ELE,
    essa: birl_1.default.ESSA,
    filho: birl_1.default.FILHO,
    frango: birl_1.default.FRANGO,
    gente: birl_1.default.GENTE,
    home: birl_1.default.HOME,
    hora: birl_1.default.HORA,
    mais: birl_1.default.MAIS,
    maluco: birl_1.default.MALUCO,
    monstrao: birl_1.default.MONSTRAO,
    monstro: birl_1.default.MONSTRO,
    monstrinho: birl_1.default.MONSTRINHO,
    nao: birl_1.default.NAO,
    negativa: birl_1.default.NEGATIVA,
    o: birl_1.default.O,
    oh: birl_1.default.OH,
    po: birl_1.default.PO,
    porra: birl_1.default.PORRA,
    puta: birl_1.default.PUTA,
    que: birl_1.default.QUE,
    quer: birl_1.default.QUER,
    quero: birl_1.default.QUERO,
    sai: birl_1.default.SAI,
    show: birl_1.default.SHOW,
    ta: birl_1.default.TA,
    vai: birl_1.default.VAI,
    vamo: birl_1.default.VAMO,
    ver: birl_1.default.VER,
    trapezio: birl_1.default.TRAPEZIO,
};
