// @designliquido/delegua/lexador/dialetos/palavras-reservadas/mapler.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.palavrasReservadas = void 0;
const mapler_1 = __importDefault(require("../../../tipos-de-simbolos/mapler"));
exports.palavrasReservadas = {
    ate: mapler_1.default.ATE,
    cadeia: mapler_1.default.CADEIA,
    caractere: mapler_1.default.CARACTERE,
    de: mapler_1.default.DE,
    e: mapler_1.default.E,
    enquanto: mapler_1.default.ENQUANTO,
    entao: mapler_1.default.ENTAO,
    escrever: mapler_1.default.ESCREVER,
    faca: mapler_1.default.FACA,
    falso: mapler_1.default.FALSO,
    fim: mapler_1.default.FIM,
    fimenquanto: mapler_1.default.FIM_ENQUANTO,
    fimpara: mapler_1.default.FIM_PARA,
    fimse: mapler_1.default.FIM_SE,
    inicio: mapler_1.default.INICIO,
    inteiro: mapler_1.default.INTEIRO,
    ler: mapler_1.default.LER,
    logico: mapler_1.default.LOGICO,
    nao: mapler_1.default.NEGACAO,
    ou: mapler_1.default.OU,
    para: mapler_1.default.PARA,
    real: mapler_1.default.REAL,
    repita: mapler_1.default.REPITA,
    se: mapler_1.default.SE,
    senao: mapler_1.default.SENAO,
    variaveis: mapler_1.default.VARIAVEIS,
    verdadeiro: mapler_1.default.VERDADEIRO,
    vetor: mapler_1.default.VETOR,
};
