// @designliquido/delegua/lexador/dialetos/palavras-reservadas/portugol-studio.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.palavrasReservadas = void 0;
const portugol_studio_1 = __importDefault(require("../../../tipos-de-simbolos/portugol-studio"));
exports.palavrasReservadas = {
    cadeia: portugol_studio_1.default.CADEIA,
    caracter: portugol_studio_1.default.CARACTER,
    caso: portugol_studio_1.default.CASO,
    const: portugol_studio_1.default.CONSTANTE,
    contrario: portugol_studio_1.default.CONTRARIO,
    enquanto: portugol_studio_1.default.ENQUANTO,
    escolha: portugol_studio_1.default.ESCOLHA,
    escreva: portugol_studio_1.default.ESCREVA,
    e: portugol_studio_1.default.E,
    faca: portugol_studio_1.default.FACA,
    falso: portugol_studio_1.default.FALSO,
    funcao: portugol_studio_1.default.FUNCAO,
    inteiro: portugol_studio_1.default.INTEIRO,
    leia: portugol_studio_1.default.LEIA,
    logico: portugol_studio_1.default.LOGICO,
    nao: portugol_studio_1.default.NEGACAO,
    ou: portugol_studio_1.default.OU,
    para: portugol_studio_1.default.PARA,
    pare: portugol_studio_1.default.PARE,
    programa: portugol_studio_1.default.PROGRAMA,
    real: portugol_studio_1.default.REAL,
    retorne: portugol_studio_1.default.RETORNE,
    se: portugol_studio_1.default.SE,
    senao: portugol_studio_1.default.SENAO,
    vazio: portugol_studio_1.default.VAZIO,
    verdadeiro: portugol_studio_1.default.VERDADEIRO,
};
