// @designliquido/potigol/lexador/palavras-reservadas.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.palavrasReservadas = void 0;
const lexico_regular_1 = __importDefault(require("../tipos-de-simbolos/lexico-regular"));
exports.palavrasReservadas = {
    ate: lexico_regular_1.default.ATE,
    até: lexico_regular_1.default.ATE,
    Caractere: lexico_regular_1.default.CARACTERE,
    caso: lexico_regular_1.default.CASO,
    de: lexico_regular_1.default.DE,
    div: lexico_regular_1.default.DIVISAO_INTEIRA,
    e: lexico_regular_1.default.E,
    escolha: lexico_regular_1.default.ESCOLHA,
    entao: lexico_regular_1.default.ENTAO,
    então: lexico_regular_1.default.ENTAO,
    enquanto: lexico_regular_1.default.ENQUANTO,
    escreva: lexico_regular_1.default.ESCREVA,
    faca: lexico_regular_1.default.FACA,
    faça: lexico_regular_1.default.FACA,
    falso: lexico_regular_1.default.FALSO,
    fim: lexico_regular_1.default.FIM,
    imprima: lexico_regular_1.default.IMPRIMA,
    Inteiro: lexico_regular_1.default.INTEIRO,
    leia_inteiro: lexico_regular_1.default.LEIA_INTEIRO,
    leia_inteiros: lexico_regular_1.default.LEIA_INTEIROS,
    leia_reais: lexico_regular_1.default.LEIA_REAIS,
    leia_real: lexico_regular_1.default.LEIA_REAL,
    leia_texto: lexico_regular_1.default.LEIA_TEXTO,
    leia_textos: lexico_regular_1.default.LEIA_TEXTOS,
    Logico: lexico_regular_1.default.LOGICO,
    Lógico: lexico_regular_1.default.LÓGICO,
    mod: lexico_regular_1.default.MODULO,
    ou: lexico_regular_1.default.OU,
    nao: lexico_regular_1.default.NEGACAO,
    não: lexico_regular_1.default.NEGACAO,
    para: lexico_regular_1.default.PARA,
    passo: lexico_regular_1.default.PASSO,
    Real: lexico_regular_1.default.REAL,
    retorne: lexico_regular_1.default.RETORNE,
    se: lexico_regular_1.default.SE,
    senao: lexico_regular_1.default.SENAO,
    senão: lexico_regular_1.default.SENAO,
    Texto: lexico_regular_1.default.TEXTO,
    tipo: lexico_regular_1.default.TIPO,
    var: lexico_regular_1.default.VARIAVEL,
    verdadeiro: lexico_regular_1.default.VERDADEIRO,
    qual_tipo: lexico_regular_1.default.QUAL_TIPO,
};
