// @designliquido/portugol-studio/lexador/palavras-reservadas.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.palavrasReservadas = void 0;
const lexico_regular_1 = __importDefault(require("../tipos-de-simbolos/lexico-regular"));
exports.palavrasReservadas = {
    cadeia: lexico_regular_1.default.CADEIA,
    caracter: lexico_regular_1.default.CARACTER,
    caso: lexico_regular_1.default.CASO,
    const: lexico_regular_1.default.CONSTANTE,
    contrario: lexico_regular_1.default.CONTRARIO,
    enquanto: lexico_regular_1.default.ENQUANTO,
    escolha: lexico_regular_1.default.ESCOLHA,
    escreva: lexico_regular_1.default.ESCREVA,
    e: lexico_regular_1.default.E,
    faca: lexico_regular_1.default.FACA,
    falso: lexico_regular_1.default.FALSO,
    funcao: lexico_regular_1.default.FUNCAO,
    inteiro: lexico_regular_1.default.INTEIRO,
    leia: lexico_regular_1.default.LEIA,
    logico: lexico_regular_1.default.LOGICO,
    nao: lexico_regular_1.default.NEGACAO,
    ou: lexico_regular_1.default.OU,
    para: lexico_regular_1.default.PARA,
    pare: lexico_regular_1.default.PARE,
    programa: lexico_regular_1.default.PROGRAMA,
    real: lexico_regular_1.default.REAL,
    retorne: lexico_regular_1.default.RETORNE,
    se: lexico_regular_1.default.SE,
    senao: lexico_regular_1.default.SENAO,
    vazio: lexico_regular_1.default.VAZIO,
    verdadeiro: lexico_regular_1.default.VERDADEIRO,
};
