// @designliquido/visualg/interpretador/comum.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.visitarExpressaoLeia = exports.visitarDeclaracaoAleatorio = exports.visitarExpressaoAtribuicaoPorIndicesMatriz = exports.visitarExpressaoAcessoElementoMatriz = exports.resolverIncrementoPara = exports.visitarExpressaoLogica = exports.visitarExpressaoBinaria = exports.atribuirVariavel = exports.visitarDeclaracaoInicioAlgoritmo = exports.visitarDeclaracaoCabecalhoPrograma = void 0;
const construtos_1 = require("@designliquido/delegua/construtos");
const declaracoes_1 = require("@designliquido/delegua/declaracoes");
const lexador_1 = require("@designliquido/delegua/lexador");
const inferenciador_1 = require("./inferenciador");
const visualg_1 = __importDefault(require("@designliquido/delegua/tipos-de-simbolos/visualg"));
const excecoes_1 = require("@designliquido/delegua/excecoes");
async function visitarDeclaracaoCabecalhoPrograma(interpretador, declaracao) {
    return Promise.resolve();
}
exports.visitarDeclaracaoCabecalhoPrograma = visitarDeclaracaoCabecalhoPrograma;
async function visitarDeclaracaoInicioAlgoritmo(interpretador, declaracao) {
    return Promise.resolve();
}
exports.visitarDeclaracaoInicioAlgoritmo = visitarDeclaracaoInicioAlgoritmo;
async function atribuirVariavel(interpretador, expressao, valor) {
    if (expressao instanceof construtos_1.Variavel) {
        interpretador.pilhaEscoposExecucao.atribuirVariavel(expressao.simbolo, valor);
        return;
    }
    if (expressao instanceof construtos_1.AcessoIndiceVariavel) {
        const promises = await Promise.all([
            interpretador.avaliar(expressao.entidadeChamada),
            interpretador.avaliar(expressao.indice),
        ]);
        let alvo = promises[0];
        let indice = promises[1];
        let valorAlvo;
        let valorIndice;
        if (alvo.hasOwnProperty('valor')) {
            valorAlvo = alvo.valor;
        }
        else {
            valorAlvo = alvo;
        }
        if (indice.hasOwnProperty('valor')) {
            valorIndice = indice.valor;
        }
        else {
            valorIndice = indice;
        }
        const subtipo = String(alvo.tipo).replace('[]', '');
        let valorResolvido;
        switch (subtipo) {
            case 'inteiro':
                valorResolvido = parseInt(valor);
                break;
            case 'lógico':
                valorResolvido = Boolean(valor);
                break;
            case 'número':
                valorResolvido = Number(valor);
                break;
            case 'texto':
                valorResolvido = String(valor);
                break;
            default:
                valorResolvido = valor;
                break;
        }
        valorAlvo[valorIndice] = valorResolvido;
    }
}
exports.atribuirVariavel = atribuirVariavel;
async function avaliar(interpretador, expressao) {
    return await expressao.aceitar(interpretador);
}
function eIgual(esquerda, direita) {
    if (esquerda === null && direita === null)
        return true;
    if (esquerda === null)
        return false;
    return esquerda === direita;
}
function eVerdadeiro(objeto) {
    if (objeto === null)
        return false;
    if (typeof objeto === 'boolean')
        return Boolean(objeto);
    if (objeto.hasOwnProperty('valor')) {
        return Boolean(objeto.valor);
    }
    return true;
}
function verificarOperandosNumeros(operador, direita, esquerda) {
    const tipoDireita = direita.tipo ? direita.tipo : typeof direita === 'number' ? 'número' : String(NaN);
    const tipoEsquerda = esquerda.tipo ? esquerda.tipo : typeof esquerda === 'number' ? 'número' : String(NaN);
    const tiposNumericos = ['inteiro', 'numero', 'número', 'real'];
    if (tiposNumericos.includes(tipoDireita.toLowerCase()) && tiposNumericos.includes(tipoEsquerda.toLowerCase()))
        return;
    throw new excecoes_1.ErroEmTempoDeExecucao(operador, 'Operadores precisam ser números.', operador.linha);
}
/**
 * Método de visita de expressão binária.
 * Reintroduzido pelas particularidades do VisuAlg.
 * @param expressao A expressão binária.
 * @returns O resultado da resolução da expressão.
 */
async function visitarExpressaoBinaria(interpretador, expressao) {
    try {
        const promises = await Promise.all([
            avaliar(interpretador, expressao.esquerda),
            avaliar(interpretador, expressao.direita),
        ]);
        const esquerda = promises[0];
        const direita = promises[1];
        let valorEsquerdo = (esquerda === null || esquerda === void 0 ? void 0 : esquerda.hasOwnProperty('valor')) ? esquerda.valor : esquerda;
        let valorDireito = (direita === null || direita === void 0 ? void 0 : direita.hasOwnProperty('valor')) ? direita.valor : direita;
        // No VisuAlg, uma variável pode resolver para função porque funções não precisam ter parênteses.
        // Esta parte evita o problema.
        if (valorEsquerdo && valorEsquerdo.hasOwnProperty('funcao')) {
            valorEsquerdo = valorEsquerdo.funcao();
        }
        if (valorDireito && valorDireito.hasOwnProperty('funcao')) {
            valorDireito = valorDireito.funcao();
        }
        const tipoEsquerdo = (esquerda === null || esquerda === void 0 ? void 0 : esquerda.hasOwnProperty('tipo')) ? esquerda.tipo : (0, inferenciador_1.inferirTipoVariavel)(esquerda);
        const tipoDireito = (direita === null || direita === void 0 ? void 0 : direita.hasOwnProperty('tipo')) ? direita.tipo : (0, inferenciador_1.inferirTipoVariavel)(direita);
        switch (expressao.operador.tipo) {
            case visualg_1.default.EXPONENCIACAO:
                verificarOperandosNumeros(expressao.operador, esquerda, direita);
                return Math.pow(valorEsquerdo, valorDireito);
            case visualg_1.default.MAIOR:
                verificarOperandosNumeros(expressao.operador, esquerda, direita);
                return Number(valorEsquerdo) > Number(valorDireito);
            case visualg_1.default.MAIOR_IGUAL:
                verificarOperandosNumeros(expressao.operador, esquerda, direita);
                return Number(valorEsquerdo) >= Number(valorDireito);
            case visualg_1.default.MENOR:
                verificarOperandosNumeros(expressao.operador, esquerda, direita);
                return Number(valorEsquerdo) < Number(valorDireito);
            case visualg_1.default.MENOR_IGUAL:
                verificarOperandosNumeros(expressao.operador, esquerda, direita);
                return Number(valorEsquerdo) <= Number(valorDireito);
            case visualg_1.default.SUBTRACAO:
                verificarOperandosNumeros(expressao.operador, esquerda, direita);
                return Number(valorEsquerdo) - Number(valorDireito);
            case visualg_1.default.ADICAO:
                let tiposNumericos = ['inteiro', 'numero', 'número', 'real'];
                if (tiposNumericos.includes(tipoEsquerdo.toLowerCase())
                    && tiposNumericos.includes(tipoDireito.toLowerCase())) {
                    return Number(valorEsquerdo) + Number(valorDireito);
                }
                else {
                    return String(valorEsquerdo) + String(valorDireito);
                }
            case visualg_1.default.DIVISAO:
                verificarOperandosNumeros(expressao.operador, esquerda, direita);
                return Number(valorEsquerdo) / Number(valorDireito);
            case visualg_1.default.DIVISAO_INTEIRA:
                verificarOperandosNumeros(expressao.operador, esquerda, direita);
                return Math.floor(Number(valorEsquerdo) / Number(valorDireito));
            case visualg_1.default.MULTIPLICACAO:
                if (tipoEsquerdo === 'texto' || tipoDireito === 'texto') {
                    // Sem ambos os valores resolvem como texto, multiplica normal.
                    // Se apenas um resolve como texto, o outro repete o
                    // texto n vezes, sendo n o valor do outro.
                    if (tipoEsquerdo === 'texto' && tipoDireito === 'texto') {
                        return Number(valorEsquerdo) * Number(valorDireito);
                    }
                    if (tipoEsquerdo === 'texto') {
                        return valorEsquerdo.repeat(Number(valorDireito));
                    }
                    return valorDireito.repeat(Number(valorEsquerdo));
                }
                return Number(valorEsquerdo) * Number(valorDireito);
            case visualg_1.default.MODULO:
                verificarOperandosNumeros(expressao.operador, esquerda, direita);
                return Number(valorEsquerdo) % Number(valorDireito);
            case visualg_1.default.DIFERENTE:
                return !eIgual(valorEsquerdo, valorDireito);
            case visualg_1.default.IGUAL:
                return eIgual(valorEsquerdo, valorDireito);
        }
    }
    catch (erro) {
        return Promise.reject(erro);
    }
}
exports.visitarExpressaoBinaria = visitarExpressaoBinaria;
async function visitarExpressaoLogica(interpretador, expressao) {
    const esquerda = await avaliar(interpretador, expressao.esquerda);
    // se um estado for verdadeiro, retorna verdadeiro
    if (expressao.operador.tipo === visualg_1.default.OU) {
        if (eVerdadeiro(esquerda))
            return esquerda;
    }
    // se a primeira variável é verdadeiro, retorna a segunda invertida
    if (expressao.operador.tipo === visualg_1.default.XOU) {
        const valorDireito = await avaliar(interpretador, expressao.direita);
        return eVerdadeiro(esquerda) !== eVerdadeiro(valorDireito);
    }
    // se um estado for falso, retorna falso
    if (expressao.operador.tipo === visualg_1.default.E) {
        if (!eVerdadeiro(esquerda))
            return esquerda;
    }
    return await avaliar(interpretador, expressao.direita);
}
exports.visitarExpressaoLogica = visitarExpressaoLogica;
/* Isso existe porque o laço `para` do VisuAlg pode ter o passo positivo ou negativo
 * dependendo dos operandos de início e fim, que só são possíveis de determinar
 * em tempo de execução.
 * Quando um dos operandos é uma variável, tanto a condição do laço quanto o
 * passo são considerados indefinidos aqui.
 */
async function resolverIncrementoPara(interpretador, declaracao) {
    if (declaracao.resolverIncrementoEmExecucao) {
        const promises = await Promise.all([
            avaliar(interpretador, declaracao.condicao.esquerda),
            avaliar(interpretador, declaracao.condicao.direita),
        ]);
        const operandoEsquerdo = promises[0];
        const operandoDireito = promises[1];
        const valorAtualEsquerdo = operandoEsquerdo.hasOwnProperty('valor') ? operandoEsquerdo.valor : operandoEsquerdo;
        const valorAtualDireito = operandoDireito.hasOwnProperty('valor') ? operandoDireito.valor : operandoDireito;
        if (valorAtualEsquerdo < valorAtualDireito) {
            declaracao.condicao.operador = new lexador_1.Simbolo(visualg_1.default.MENOR_IGUAL, '', '', Number(declaracao.linha), declaracao.hashArquivo);
            declaracao.incrementar.condicaoPara.operador = new lexador_1.Simbolo(visualg_1.default.MENOR, '', '', Number(declaracao.linha), declaracao.hashArquivo);
            declaracao.incrementar.incremento.expressao.valor.direita = new construtos_1.Literal(declaracao.hashArquivo, Number(declaracao.linha), 1);
        }
        else {
            declaracao.condicao.operador = new lexador_1.Simbolo(visualg_1.default.MAIOR_IGUAL, '', '', Number(declaracao.linha), declaracao.hashArquivo);
            declaracao.incrementar.condicaoPara.operador = new lexador_1.Simbolo(visualg_1.default.MAIOR, '', '', Number(declaracao.linha), declaracao.hashArquivo);
            declaracao.incrementar.incremento.expressao.valor.direita = new construtos_1.Unario(declaracao.hashArquivo, new lexador_1.Simbolo(visualg_1.default.SUBTRACAO, '-', undefined, declaracao.linha, declaracao.hashArquivo), new construtos_1.Literal(declaracao.hashArquivo, Number(declaracao.linha), 1), 'ANTES');
        }
    }
}
exports.resolverIncrementoPara = resolverIncrementoPara;
async function visitarExpressaoAcessoElementoMatriz(interpretador, expressao) {
    const promises = await Promise.all([
        avaliar(interpretador, expressao.entidadeChamada),
        avaliar(interpretador, expressao.indicePrimario),
        avaliar(interpretador, expressao.indiceSecundario),
    ]);
    const variavelObjeto = promises[0];
    const indicePrimario = promises[1];
    const indiceSecundario = promises[2];
    const objeto = variavelObjeto.hasOwnProperty('valor') ? variavelObjeto.valor : variavelObjeto;
    let valorIndicePrimario = indicePrimario.hasOwnProperty('valor') ? indicePrimario.valor : indicePrimario;
    let valorIndiceSecundario = indiceSecundario.hasOwnProperty('valor') ? indiceSecundario.valor : indiceSecundario;
    if (Array.isArray(objeto)) {
        if (!Number.isInteger(valorIndicePrimario) || !Number.isInteger(valorIndiceSecundario)) {
            return Promise.reject(new excecoes_1.ErroEmTempoDeExecucao(expressao.simboloFechamento, 'Somente inteiros podem ser usados para indexar um vetor.', expressao.linha));
        }
        if (valorIndicePrimario < 0 && objeto.length !== 0) {
            while (valorIndicePrimario < 0) {
                valorIndicePrimario += objeto.length;
            }
        }
        if (valorIndiceSecundario < 0 && objeto.length !== 0) {
            while (valorIndiceSecundario < 0) {
                valorIndiceSecundario += objeto.length;
            }
        }
        if (valorIndicePrimario >= objeto.length || valorIndiceSecundario >= objeto.length) {
            return Promise.reject(new excecoes_1.ErroEmTempoDeExecucao(expressao.simboloFechamento, 'Índice do vetor fora do intervalo.', expressao.linha));
        }
        return objeto[valorIndicePrimario][valorIndiceSecundario];
    }
    return Promise.reject(new excecoes_1.ErroEmTempoDeExecucao(expressao.entidadeChamada.valor, 'Somente listas, dicionários, classes e objetos podem ser mudados por sobrescrita.', expressao.linha));
}
exports.visitarExpressaoAcessoElementoMatriz = visitarExpressaoAcessoElementoMatriz;
async function visitarExpressaoAtribuicaoPorIndicesMatriz(interpretador, expressao) {
    const promises = await Promise.all([
        avaliar(interpretador, expressao.objeto),
        avaliar(interpretador, expressao.indicePrimario),
        avaliar(interpretador, expressao.indiceSecundario),
        avaliar(interpretador, expressao.valor),
    ]);
    let objeto = promises[0];
    let indicePrimario = promises[1];
    let indiceSecundario = promises[2];
    const valor = promises[3];
    objeto = objeto.hasOwnProperty('valor') ? objeto.valor : objeto;
    indicePrimario = indicePrimario.hasOwnProperty('valor') ? indicePrimario.valor : indicePrimario;
    indiceSecundario = indiceSecundario.hasOwnProperty('valor') ? indiceSecundario.valor : indiceSecundario;
    if (Array.isArray(objeto)) {
        if (indicePrimario < 0 && objeto.length !== 0) {
            while (indicePrimario < 0) {
                indicePrimario += objeto.length;
            }
        }
        if (indiceSecundario < 0 && objeto.length !== 0) {
            while (indiceSecundario < 0) {
                indiceSecundario += objeto.length;
            }
        }
        while (objeto.length < indicePrimario || objeto.length < indiceSecundario) {
            objeto.push(null);
        }
        objeto[indicePrimario][indiceSecundario] = valor;
        return Promise.resolve();
    }
    return Promise.reject(new excecoes_1.ErroEmTempoDeExecucao(expressao.objeto.nome, 'Somente listas, dicionários, classes e objetos podem ser mudados por sobrescrita.', expressao.linha));
}
exports.visitarExpressaoAtribuicaoPorIndicesMatriz = visitarExpressaoAtribuicaoPorIndicesMatriz;
async function encontrarLeiaNoAleatorio(interpretador, declaracao, menorNumero, maiorNumero) {
    if ('declaracoes' in declaracao) {
        // Se a declaração tiver um campo 'declaracoes', ela é um Bloco
        const declaracoes = declaracao.declaracoes;
        for (const subDeclaracao of declaracoes) {
            encontrarLeiaNoAleatorio(interpretador, subDeclaracao, menorNumero, maiorNumero);
        }
    }
    else if (declaracao instanceof declaracoes_1.Leia) {
        // Se encontrarmos um Leia, podemos efetuar as operações imediatamente
        for (const argumento of declaracao.argumentos) {
            const arg1 = await interpretador.avaliar(argumento);
            const tipoDe = arg1.tipo || (0, inferenciador_1.inferirTipoVariavel)(arg1);
            const valor = tipoDe === 'texto' ? palavraAleatoriaCom5Digitos() : gerarNumeroAleatorio(menorNumero, maiorNumero);
            atribuirVariavel(interpretador, argumento, valor);
        }
    }
}
function gerarNumeroAleatorio(min, max) {
    return Math.floor(Math.random() * (max - min) + min);
}
function palavraAleatoriaCom5Digitos() {
    const caracteres = 'abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ';
    let palavra = '';
    for (let i = 0; i < 5; i++) {
        const indiceAleatorio = Math.floor(Math.random() * caracteres.length);
        palavra += caracteres.charAt(indiceAleatorio);
    }
    return palavra;
}
async function visitarDeclaracaoAleatorio(interpretador, expressao) {
    let retornoExecucao;
    try {
        let menorNumero = 0;
        let maiorNumero = 100;
        if (expressao.argumentos) {
            menorNumero = Math.min(expressao.argumentos.min, expressao.argumentos.max);
            maiorNumero = Math.max(expressao.argumentos.min, expressao.argumentos.max);
        }
        for (let corpoDeclaracao of expressao.corpo.declaracoes) {
            encontrarLeiaNoAleatorio(interpretador, corpoDeclaracao, menorNumero, maiorNumero);
            retornoExecucao = await interpretador.executar(corpoDeclaracao);
        }
    }
    catch (error) {
        interpretador.erros.push({
            erroInterno: error,
            linha: expressao.linha,
            hashArquivo: expressao.hashArquivo,
        });
        return Promise.reject(error);
    }
    return retornoExecucao;
}
exports.visitarDeclaracaoAleatorio = visitarDeclaracaoAleatorio;
async function visitarExpressaoLeia(interpretador, expressao, mensagemPrompt) {
    // Verifica se a leitura deve ser interrompida antes de prosseguir
    if (!expressao.eParaInterromper) {
        for (let argumento of expressao.argumentos) {
            const promessaLeitura = () => new Promise((resolucao) => interpretador.interfaceEntradaSaida.question(mensagemPrompt, (resposta) => {
                mensagemPrompt = '> ';
                resolucao(resposta);
            }));
            const valorLido = await promessaLeitura();
            await atribuirVariavel(interpretador, argumento, valorLido);
        }
    }
}
exports.visitarExpressaoLeia = visitarExpressaoLeia;
