// @designliquido/birl/lexador/palavras-reservadas.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.palavrasReservadas = void 0;
const lexico_regular_1 = __importDefault(require("../tipos-de-simbolos/lexico-regular"));
exports.palavrasReservadas = {
    a: lexico_regular_1.default.A,
    ai: lexico_regular_1.default.AI,
    ajuda: lexico_regular_1.default.AJUDA,
    bambam: lexico_regular_1.default.BAMBAM,
    birl: lexico_regular_1.default.BIRL,
    bora: lexico_regular_1.default.BORA,
    biceps: lexico_regular_1.default.BICEPS,
    ce: lexico_regular_1.default.CE,
    cumpade: lexico_regular_1.default.CUMPADE,
    da: lexico_regular_1.default.DA,
    dar: lexico_regular_1.default.DAR,
    do: lexico_regular_1.default.DO,
    doente: lexico_regular_1.default.DOENTE,
    descendente: lexico_regular_1.default.DESCENDENTE,
    ele: lexico_regular_1.default.ELE,
    essa: lexico_regular_1.default.ESSA,
    filho: lexico_regular_1.default.FILHO,
    frango: lexico_regular_1.default.FRANGO,
    gente: lexico_regular_1.default.GENTE,
    home: lexico_regular_1.default.HOME,
    hora: lexico_regular_1.default.HORA,
    mais: lexico_regular_1.default.MAIS,
    maluco: lexico_regular_1.default.MALUCO,
    monstrao: lexico_regular_1.default.MONSTRAO,
    monstro: lexico_regular_1.default.MONSTRO,
    monstrinho: lexico_regular_1.default.MONSTRINHO, // Essa palavra reservada faz referencia a um short int que em C tem 2bytes ou 16 bits, mas não tratamos ela assim, aqui ela é ignorada e a tratamos como um inteiros de 4 bytes ou 32 bits.
    nao: lexico_regular_1.default.NAO,
    negativa: lexico_regular_1.default.NEGATIVA,
    o: lexico_regular_1.default.O,
    oh: lexico_regular_1.default.OH,
    po: lexico_regular_1.default.PO,
    porra: lexico_regular_1.default.PORRA,
    puta: lexico_regular_1.default.PUTA,
    que: lexico_regular_1.default.QUE,
    quer: lexico_regular_1.default.QUER,
    quero: lexico_regular_1.default.QUERO,
    sai: lexico_regular_1.default.SAI,
    show: lexico_regular_1.default.SHOW,
    ta: lexico_regular_1.default.TA,
    vai: lexico_regular_1.default.VAI,
    vamo: lexico_regular_1.default.VAMO,
    ver: lexico_regular_1.default.VER,
    trapezio: lexico_regular_1.default.TRAPEZIO,
};
